src/catalog/postgres/pg_proc_impl.cpp

```diff
@@ -0,0 +1,791 @@
+#include "catalog/postgres/pg_proc_impl.h"
+
+#include <vector>
+
+#include "catalog/database_catalog.h"
+#include "catalog/postgres/builder.h"
+#include "catalog/postgres/pg_namespace.h"
+#include "catalog/postgres/pg_proc.h"
+#include "common/error/error_code.h"
+#include "common/error/exception.h"
+#include "execution/functions/function_context.h"
+#include "storage/index/index.h"
+#include "storage/sql_table.h"
+#include "transaction/deferred_action_manager.h"
+
+namespace noisepage::catalog::postgres {
+
+PgProcImpl::PgProcImpl(db_oid_t db_oid) : db_oid_(db_oid) {}
+
+void PgProcImpl::BootstrapPRIs() {
+  const std::vector<col_oid_t> pg_proc_all_oids{PgProc::PG_PRO_ALL_COL_OIDS.cbegin(),
+                                                PgProc::PG_PRO_ALL_COL_OIDS.cend()};
+  pg_proc_all_cols_pri_ = procs_->InitializerForProjectedRow(pg_proc_all_oids);
+  pg_proc_all_cols_prm_ = procs_->ProjectionMapForOids(pg_proc_all_oids);
+
+  const std::vector<col_oid_t> set_pg_proc_ptr_oids{PgProc::PRO_CTX_PTR_COL_OID};
+  pg_proc_ptr_pri_ = procs_->InitializerForProjectedRow(set_pg_proc_ptr_oids);
+}
+
+void PgProcImpl::Bootstrap(const common::ManagedPointer<DatabaseCatalog> dbc,
+                           const common::ManagedPointer<transaction::TransactionContext> txn) {
+  UNUSED_ATTRIBUTE bool retval;
+
+  retval = dbc->CreateTableEntry(txn, postgres::PgProc::PRO_TABLE_OID, postgres::NAMESPACE_CATALOG_NAMESPACE_OID,
+                                 "pg_proc", postgres::Builder::GetProcTableSchema());
+  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
+  retval = dbc->SetTablePointer(txn, postgres::PgProc::PRO_TABLE_OID, procs_);
+  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
+
+  retval = dbc->CreateIndexEntry(txn, postgres::NAMESPACE_CATALOG_NAMESPACE_OID, postgres::PgProc::PRO_TABLE_OID,
+                                 postgres::PgProc::PRO_OID_INDEX_OID, "pg_proc_oid_index",
+                                 postgres::Builder::GetProcOidIndexSchema(db_oid_));
+  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
+  retval = dbc->SetIndexPointer(txn, postgres::PgProc::PRO_OID_INDEX_OID, procs_oid_index_);
+  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
+
+  retval = dbc->CreateIndexEntry(txn, postgres::NAMESPACE_CATALOG_NAMESPACE_OID, postgres::PgProc::PRO_TABLE_OID,
+                                 postgres::PgProc::PRO_NAME_INDEX_OID, "pg_proc_name_index",
+                                 postgres::Builder::GetProcNameIndexSchema(db_oid_));
+  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
+  retval = dbc->SetIndexPointer(txn, postgres::PgProc::PRO_NAME_INDEX_OID, procs_name_index_);
+  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
+
+  BootstrapProcs(dbc, txn);
+}
+
+std::vector<execution::functions::FunctionContext *> PgProcImpl::TearDownGetFuncContexts(
```

Comment:
See comment in `database_catalog.cpp` about whether this should actually return the function that deletes underlying objects instead.

Reply:
Done.